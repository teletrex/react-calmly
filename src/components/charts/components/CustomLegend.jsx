/* <LICENSE>
*
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
*
* </LICENSE>
*  */




import React, { useCallback, useMemo, useState } from 'react';
import PropTypes from 'prop-types';
import  settings from '../../../settings';
import { ChevronDown, ChevronUp} from '@carbon/icons-react';
import classNames from 'classnames';
import uniqueId from 'lodash/uniqueId';
import noop from 'lodash/noop';

import { useTranslation } from '../../../translation';
import Checkbox from '../../checkbox';

const { prefix } = settings;

export const getTitleClassName = collapsable =>
  classNames(`${prefix}--recharts__legend_title`, {
    [`${prefix}--recharts__legend_title-collapsable`]: collapsable,
  });

export const getIconWhen = collapsed => (collapsed ? <ChevronDown size={16} /> : <ChevronUp size={16} />);

const CustomLegend = ({
  data,
  onChange,
  withMetrics,
  collapsable,
  collapsed,
  onCollapse,
  selectAllEntriesAfterAllWereDeselected,
}) => {
  const { t } = useTranslation();

  const [hiddenEntries, setHiddenEntries] = useState([]);

  const handleLegendItemClick = useCallback(
    entry => checked => {
      let newHiddenEntries;

      const isAllDeselected = !checked && hiddenEntries.length === data.length - 1;

      if (selectAllEntriesAfterAllWereDeselected && isAllDeselected) {
        newHiddenEntries = [];
      } else {
        newHiddenEntries = checked
          ? hiddenEntries.filter(entryName => entryName !== entry.name)
          : [...hiddenEntries, entry.name];
      }

      setHiddenEntries(newHiddenEntries);
      onChange(newHiddenEntries);
    },
    [data.length, hiddenEntries, onChange, selectAllEntriesAfterAllWereDeselected]
  );

  const isChecked = entry => !hiddenEntries.includes(entry.name);

  const className = classNames(`${prefix}--recharts__legend`, {
    [`${prefix}--recharts__legend-with-metrics`]: withMetrics,
    [`${prefix}--recharts__legend-collapsed`]: collapsed,
  });

  const collapseArrow = useMemo(() => getIconWhen(collapsed), [collapsed]);
  const dataA = [];
  const dataB = [];

  data.forEach(it => {
    if (it.comparingGroup) {
      dataA.push(it);
    } else {
      dataB.push(it);
    }
  });

  const legendRenderItems = (data, id) =>
    data.map((entry, index) => (
      <li
        key={`leg-${index}-${id}`}
        style={{
          '--checkbox-background-color': entry.color,
          '--checkbox-border-color': entry.color,
        }}
      >
        <Checkbox
          checked={isChecked(entry)}
          data-testid={`checkbox-${entry.name}`}
          id={uniqueId(`leg-${entry.name}`)}
          labelText={entry.name}
          onChange={handleLegendItemClick(entry)}
        />
      </li>
    ));
  return (
    <div className={className} data-testid="recharts__legend">
      <button
        className={getTitleClassName(collapsable)}
        data-testid="legend-title"
        onClick={onCollapse}
        type="button"
      >
        <span>{t('Legend')}</span>
        {collapsable && collapseArrow}
      </button>
      {!collapsed && (
        <ul data-testid="legend-items">
          {legendRenderItems(dataA, 'A')}
          {dataA.length > 0 && dataB.length > 0 && (
            <li>
              <hr />
            </li>
          )}
          {legendRenderItems(dataB, 'B')}
        </ul>
      )}
    </div>
  );
};

CustomLegend.defaultProps = {
  collapsable: false,
  collapsed: false,
  data: [],
  onCollapse: noop,
  selectAllEntriesAfterAllWereDeselected: true,
  withMetrics: false,
};

CustomLegend.propTypes = {
  collapsable: PropTypes.bool,
  collapsed: PropTypes.bool,
  data: PropTypes.arrayOf(PropTypes.shape({})),
  onChange: PropTypes.func.isRequired,
  onCollapse: PropTypes.func,
  selectAllEntriesAfterAllWereDeselected: PropTypes.bool,
  withMetrics: PropTypes.bool,
};

export default CustomLegend;
