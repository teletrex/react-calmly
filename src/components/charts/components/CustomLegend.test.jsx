/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import CustomLegend from './CustomLegend';

const data = [
  { name: '2016 Q1', value: 98, color: '#29ab87' },
  { name: '2016 Q2', value: 96, color: '#0292d4' },
  { name: '2016 Q3', value: 94, color: '#d846bf' },
  { name: '2016 Q4', value: 93, color: '#f45d20' },
  { name: '2017 Q1', value: 90, color: '#f3ba00' },
  { name: '2017 Q2', value: 88, color: '#ff465a' },
  { name: '2017 Q3', value: 87, color: '#FF8C7B' },
  { name: '2017 Q4', value: 86, color: '#29ab87' },
  { name: '2018 Q1', value: 84, color: '#0292d4' },
  { name: '2018 Q2', value: 80, color: '#d846bf' },
  { name: '2018 Q3', value: 77, color: '#f45d20' },
  { name: '2018 Q4', value: 75, color: '#f3ba00' },
  { name: '2019 Q1', value: 74, color: '#ff465a' },
  { name: '2019 Q2', value: 72, color: '#FF8C7B' },
];

describe('CustomLegend', () => {
  test('verify if created', () => {
    const { container } = render(<CustomLegend data={data} onChange={() => {}} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('Uses correct data name', () => {
    const { getByText } = render(<CustomLegend data={data} onChange={() => {}} />);
    expect(getByText('2017 Q3')).not.toBeFalsy();
  });

  test('Adds the right class if `withMetrics` is true', () => {
    const { getByTestId } = render(<CustomLegend data={data} onChange={() => {}} withMetrics />);
    expect(getByTestId('recharts__legend')).toHaveClass('bx--recharts__legend-with-metrics');
  });

  test('Adds the right class if `collapsed` is true', () => {
    const { getByTestId } = render(<CustomLegend collapsed data={data} onChange={() => {}} />);
    expect(getByTestId('recharts__legend')).toHaveClass('bx--recharts__legend-collapsed');
  });

  test('Adds the right class if title is collapsable', () => {
    const { getByTestId } = render(<CustomLegend collapsable data={data} onChange={() => {}} />);
    expect(getByTestId('legend-title')).toHaveClass('bx--recharts__legend_title-collapsable');
  });

  test('verify if values are toggled correctly', () => {
    const props = { onChange: jest.fn(), data };

    const container = render(<CustomLegend {...props} />);

    fireEvent.click(container.getByTestId(`checkbox-2016 Q1`));
    fireEvent.click(container.getByTestId(`checkbox-2016 Q2`));
    fireEvent.click(container.getByTestId(`checkbox-2016 Q2`));

    expect(props.onChange).toHaveBeenCalledTimes(3);

    const {
      mock: { calls },
    } = props.onChange;

    const lastCall = calls.pop();

    expect(lastCall).toEqual([['2016 Q1']]);
  });

  test('calls onCollapse if title is clicked', () => {
    const props = { onCollapse: jest.fn(), data, onChange: () => {}, collapsable: true };
    const container = render(<CustomLegend {...props} />);

    fireEvent.click(container.getByTestId('legend-title'));
    expect(props.onCollapse).toHaveBeenCalledTimes(1);
  });

  test('should select all entries if `selectAllEntriesAfterAllUnselected` is true and all entries unselected', () => {
    const props = { onChange: jest.fn(), data, selectAllEntriesAfterAllWereDeselected: true };

    const container = render(<CustomLegend {...props} />);

    const checkboxes = container.getAllByTestId(/checkbox/);

    checkboxes.forEach(checkbox => fireEvent.click(checkbox)); // uncheck all checboxes

    expect(props.onChange).toHaveBeenCalledTimes(14);

    const {
      mock: { calls },
    } = props.onChange;

    const lastCall = calls.pop();

    expect(lastCall).toEqual([[]]);
  });
});
