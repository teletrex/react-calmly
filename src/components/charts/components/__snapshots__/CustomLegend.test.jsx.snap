// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomLegend verify if created 1`] = `
<div
  class="bx--recharts__legend"
  data-testid="recharts__legend"
>
  <button
    class="bx--recharts__legend_title"
    data-testid="legend-title"
    type="button"
  >
    <span>
      Legend
    </span>
  </button>
  <ul
    data-testid="legend-items"
  >
    <li>
      <div
        class="bx--form-item bx--checkbox-wrapper"
      >
        <input
          checked=""
          class="bx--checkbox"
          data-testid="checkbox-2016 Q1"
          id="leg-2016 Q11"
          type="checkbox"
        />
        <label
          class="bx--checkbox-label"
          for="leg-2016 Q11"
        >
          <span
            class="bx--checkbox-label-text"
          >
            2016 Q1
          </span>
        </label>
      </div>
    </li>
    <li>
      <div
        class="bx--form-item bx--checkbox-wrapper"
      >
        <input
          checked=""
          class="bx--checkbox"
          data-testid="checkbox-2016 Q2"
          id="leg-2016 Q22"
          type="checkbox"
        />
        <label
          class="bx--checkbox-label"
          for="leg-2016 Q22"
        >
          <span
            class="bx--checkbox-label-text"
          >
            2016 Q2
          </span>
        </label>
      </div>
    </li>
    <li>
      <div
        class="bx--form-item bx--checkbox-wrapper"
      >
        <input
          checked=""
          class="bx--checkbox"
          data-testid="checkbox-2016 Q3"
          id="leg-2016 Q33"
          type="checkbox"
        />
        <label
          class="bx--checkbox-label"
          for="leg-2016 Q33"
        >
          <span
            class="bx--checkbox-label-text"
          >
            2016 Q3
          </span>
        </label>
      </div>
    </li>
    <li>
      <div
        class="bx--form-item bx--checkbox-wrapper"
      >
        <input
          checked=""
          class="bx--checkbox"
          data-testid="checkbox-2016 Q4"
          id="leg-2016 Q44"
          type="checkbox"
        />
        <label
          class="bx--checkbox-label"
          for="leg-2016 Q44"
        >
          <span
            class="bx--checkbox-label-text"
          >
            2016 Q4
          </span>
        </label>
      </div>
    </li>
    <li>
      <div
        class="bx--form-item bx--checkbox-wrapper"
      >
        <input
          checked=""
          class="bx--checkbox"
          data-testid="checkbox-2017 Q1"
          id="leg-2017 Q15"
          type="checkbox"
        />
        <label
          class="bx--checkbox-label"
          for="leg-2017 Q15"
        >
          <span
            class="bx--checkbox-label-text"
          >
            2017 Q1
          </span>
        </label>
      </div>
    </li>
    <li>
      <div
        class="bx--form-item bx--checkbox-wrapper"
      >
        <input
          checked=""
          class="bx--checkbox"
          data-testid="checkbox-2017 Q2"
          id="leg-2017 Q26"
          type="checkbox"
        />
        <label
          class="bx--checkbox-label"
          for="leg-2017 Q26"
        >
          <span
            class="bx--checkbox-label-text"
          >
            2017 Q2
          </span>
        </label>
      </div>
    </li>
    <li>
      <div
        class="bx--form-item bx--checkbox-wrapper"
      >
        <input
          checked=""
          class="bx--checkbox"
          data-testid="checkbox-2017 Q3"
          id="leg-2017 Q37"
          type="checkbox"
        />
        <label
          class="bx--checkbox-label"
          for="leg-2017 Q37"
        >
          <span
            class="bx--checkbox-label-text"
          >
            2017 Q3
          </span>
        </label>
      </div>
    </li>
    <li>
      <div
        class="bx--form-item bx--checkbox-wrapper"
      >
        <input
          checked=""
          class="bx--checkbox"
          data-testid="checkbox-2017 Q4"
          id="leg-2017 Q48"
          type="checkbox"
        />
        <label
          class="bx--checkbox-label"
          for="leg-2017 Q48"
        >
          <span
            class="bx--checkbox-label-text"
          >
            2017 Q4
          </span>
        </label>
      </div>
    </li>
    <li>
      <div
        class="bx--form-item bx--checkbox-wrapper"
      >
        <input
          checked=""
          class="bx--checkbox"
          data-testid="checkbox-2018 Q1"
          id="leg-2018 Q19"
          type="checkbox"
        />
        <label
          class="bx--checkbox-label"
          for="leg-2018 Q19"
        >
          <span
            class="bx--checkbox-label-text"
          >
            2018 Q1
          </span>
        </label>
      </div>
    </li>
    <li>
      <div
        class="bx--form-item bx--checkbox-wrapper"
      >
        <input
          checked=""
          class="bx--checkbox"
          data-testid="checkbox-2018 Q2"
          id="leg-2018 Q210"
          type="checkbox"
        />
        <label
          class="bx--checkbox-label"
          for="leg-2018 Q210"
        >
          <span
            class="bx--checkbox-label-text"
          >
            2018 Q2
          </span>
        </label>
      </div>
    </li>
    <li>
      <div
        class="bx--form-item bx--checkbox-wrapper"
      >
        <input
          checked=""
          class="bx--checkbox"
          data-testid="checkbox-2018 Q3"
          id="leg-2018 Q311"
          type="checkbox"
        />
        <label
          class="bx--checkbox-label"
          for="leg-2018 Q311"
        >
          <span
            class="bx--checkbox-label-text"
          >
            2018 Q3
          </span>
        </label>
      </div>
    </li>
    <li>
      <div
        class="bx--form-item bx--checkbox-wrapper"
      >
        <input
          checked=""
          class="bx--checkbox"
          data-testid="checkbox-2018 Q4"
          id="leg-2018 Q412"
          type="checkbox"
        />
        <label
          class="bx--checkbox-label"
          for="leg-2018 Q412"
        >
          <span
            class="bx--checkbox-label-text"
          >
            2018 Q4
          </span>
        </label>
      </div>
    </li>
    <li>
      <div
        class="bx--form-item bx--checkbox-wrapper"
      >
        <input
          checked=""
          class="bx--checkbox"
          data-testid="checkbox-2019 Q1"
          id="leg-2019 Q113"
          type="checkbox"
        />
        <label
          class="bx--checkbox-label"
          for="leg-2019 Q113"
        >
          <span
            class="bx--checkbox-label-text"
          >
            2019 Q1
          </span>
        </label>
      </div>
    </li>
    <li>
      <div
        class="bx--form-item bx--checkbox-wrapper"
      >
        <input
          checked=""
          class="bx--checkbox"
          data-testid="checkbox-2019 Q2"
          id="leg-2019 Q214"
          type="checkbox"
        />
        <label
          class="bx--checkbox-label"
          for="leg-2019 Q214"
        >
          <span
            class="bx--checkbox-label-text"
          >
            2019 Q2
          </span>
        </label>
      </div>
    </li>
  </ul>
</div>
`;
